File: accelerometer/matlab/ArduinoReader.m
% This class is used to read values dumped by the MPU6050 accelerometer
% on to the Arduino serial port.
classdef ArduinoReader < handle
    properties
        arduino_port;
        port;
        baud_rate;
        
        quaternian;
        acceleration;
        gravity;
        velocity;
        displacement;

        calibration;
    end

    methods
        function [obj] = ArduinoReader(port, baud_rate)
            obj.port = port;
            obj.baud_rate = baud_rate;
            obj.quaternian = [0 0 0 0]';
            obj.acceleration = [0 0 0]';
            obj.gravity = [0 0 9.81]';
            obj.velocity = [0 0 0]';
            obj.displacement = [0 0 0]';
            obj.calibration = [1 0 1 0 1 0]';
            
            % Open port used by Arduino
            obj.arduino_port = serial(obj.port, 'BaudRate', obj.baud_rate, 'BytesAvailableFcn', {@readCallback,obj});
            obj.arduino_port.BytesAvailableFcnCount = 22;
            obj.arduino_port.BytesAvailableFcnMode = 'byte';

            % Define callback function
            function readCallback(obj, event, arduino_reader)
                if (get(obj, 'BytesAvailable') ~= 0)
                    arduino_reader.quaternian = fread(obj, 4, 'float');
                    arduino_reader.acceleration = fread(obj, 3, 'int16');
                end
            end
            
            % Start reading from Arduino port
            fopen(obj.arduino_port);
            pause(2);
            fwrite(obj.arduino_port, 1);
        end

        function [] = close(obj)
            fclose(obj.arduino_port);
        end

        function [q, a] = raw_read(obj)
            q = obj.quaternian;
            a = obj.acceleration ./ 1000;
        end

        function [R, a_real] = read(obj)
            [q, a] = obj.raw_read();
            R = quaternion_to_r(q);
            g = R'*obj.gravity;
            a_real = zeros(3, 1);
            a_real(1) = a(1)*obj.calibration(1) + obj.calibration(2) - g(1);
            a_real(2) = a(2)*obj.calibration(3) + obj.calibration(4) - g(2);
            a_real(3) = a(3)*obj.calibration(5) + obj.calibration(6) - g(3);
        end

        % Calibration function
        % Calibrate accelerometer readings
        % TODO:Calculate sample rate
        function [] = calibrate(obj)
            N = 100;
            delta = 0.5;

            sample_num = 0;
            M = zeros(3*N, 6);
            y = zeros(3*N, 1);
            lamda = 0.1;
            while sample_num < N
                [q,a] = obj.raw_read;
                magnitude = sqrt(sum(a.^2));
                
                if (magnitude > (obj.gravity(3)-delta) && magnitude < (obj.gravity(3)+delta))
                    R = quaternion_to_r(q);
                    g = R'*obj.gravity;
                    M(3*sample_num+1, 1) = a(1);
                    M(3*sample_num+1, 2) = 1;
                    M(3*sample_num+2, 3) = a(2);
                    M(3*sample_num+2, 4) = 1;
                    M(3*sample_num+3, 5) = a(3);
                    M(3*sample_num+3, 6) = 1;
                    y(3*sample_num+1:3*sample_num+3,1) = g;
                    sample_num = sample_num + 1;
                end
                
                pause(0.1);
            end

            % Calibrate using least sum of squares with penalization
            obj.calibration = pinv(M' * M + lamda * eye(6))*M' * y;
        end
    end

end

File: accelerometer/python/ArduinoReader.py
# This class is used to read values dumped by the MPU6050 accelerometer
# on to the Arduino serial port.
import numpy as np
import serial 
import time
import struct
import threading
import converters

class ArduinoReader:

    # Constructor
    def __init__(self, port, baud_rate):
        # Serial port configuration
        self.port = port
        self.baud_rate = baud_rate
        
        # Raw measurements
        self.quaternian = np.matrix('0.0; 0.0; 0.0; 0.0')
        self.acceleration = np.matrix('0.0; 0.0; 0.0')
        
        # Accelerometer related values
        self.sensitivity = 1
        self.gravity_compensation = np.matrix('9.81 -9.81; 9.81 -9.81; 9.81 -9.81')
        self.velocity = np.matrix('0.0; 0.0; 0.0')
        self.displacement = np.matrix('0.0; 0.0; 0.0')
        self.calibration = np.matrix('1.0; 0.0; 1.0; 0.0; 1.0; 0.0')

        # Threads
        self.done_reading = False

        # Open port used by Arduino
        self.arduino_port = serial.Serial(self.port, self.baud_rate)
        self.arduino_port.flush()
        #self.arduino_port.BytesAvailableFcnCount = 22;
        #self.arduino_port.BytesAvailableFcnMode = 'byte';

        # Start reading from Arduino port
        #self.arduino_port.open()
        time.sleep(2)
        #self.arduino_port.write('1')

        self.callback_thread = threading.Thread(target=readCallback, args=(self,))
        self.callback_thread.start()

    def close(self):
        self.done_reading = True
        self.callback_thread.join()
        self.arduino_port.close()

    def raw_read(self):
        q = self.quaternian;
        a = np.divide(self.acceleration, 1000.0);
        return q, a

    def read(self):
        # Read raw values coming from accelerometer
        q, a_raw = self.raw_read()

        # Convert from quaternion to rotation matrix
        R = converters.quaternion_to_r(q)

        # Calculate how much to compensate gravity by
        gravity_compensation = np.zeros((3, 1))
        for axis in range(0, 3):
            # Each axis corresponds to x,y,z
            if R[2][axis] > 0:
                # Gravity alignes with +ve axis
                gravity_compensation[axis] = R[2, axis] * self.gravity_compensation[axis, 0]
            else:
                # Gravity alignes with -ve axis
                gravity_compensation[axis] = -1 * R[2, axis] * self.gravity_compensation[axis, 1]
        
        # Calculate acceleration with gravity compensation    
        a_real = np.multiply(self.sensitivity, a_raw - gravity_compensation)
        
        return R, a_real

    # Calibration function
    # Calibrate accelerometer readings
    def calibrate(self, calibration_file):
        # Get offset constants from file
        self.sensitivity = 1
        x = 0
        y = 1
        z = 2
        positive = 0
        negative = 1

        self.gravity_compensation[x, positive] = 8.14
        self.gravity_compensation[x, negative] = -7.47
        self.gravity_compensation[y, positive] = 7.72
        self.gravity_compensation[y, negative] = -7.94
        self.gravity_compensation[z, positive] = 10.83
        self.gravity_compensation[z, negative] = -5.7
        
# This function will keep on reading from the serial port until close() is called.
# Call using a dedicated thread.
def readCallback(ar):
    while(not ar.done_reading):
        if(ar.arduino_port.in_waiting > 22):
            ar.quaternian[0] = struct.unpack('<f', ar.arduino_port.read(4))
            ar.quaternian[1] = struct.unpack('<f', ar.arduino_port.read(4))
            ar.quaternian[2] = struct.unpack('<f', ar.arduino_port.read(4))
            ar.quaternian[3] = struct.unpack('<f', ar.arduino_port.read(4))
            ar.acceleration[0] = struct.unpack('<h', ar.arduino_port.read(2))
            ar.acceleration[1] = struct.unpack('<h', ar.arduino_port.read(2))
            ar.acceleration[2] = struct.unpack('<h', ar.arduino_port.read(2))

